### PR TITLE
Escaping n translation

### PR DESCRIPTION
**Pull request by dashaluna**
Let's try again. This PR is for addition of escaping and translating functions. And minor formatting lining up.

### CODE REVIEW

**Comment by dashaluna** [comment]
@tcrsavage trying again.. This time smaller dedicated PR :) Could you check the whole plugin file that I haven't missed any escape and translation functions. Anything that could be improved further?

File: better-font-awesome.php [hunk 1]
```diff
@@ -513,18 +543,67 @@ public function text_callback( $args ) {
...
+		    <br/><br/>
+
+		    <i class="icon-coffee fa fa-coffee icon-2x fa-2x icon-rotate-90 fa-rotate-90"></i> <code>[icon name="coffee" class="icon-2x icon-rotate-90"]</code>
+		    <?php echo $code_alternative_str; ?>
+		    <code>&lt;i class="icon-coffee icon-2x icon-rotate-90"&gt;&lt;/i&gt;</code>
+
+	    </div>
+
+	    <?php
+	    return ob_get_clean();
```
Comment (dashaluna) [inline better-font-awesome.php]: @tcrsavage do you think I should add a filter to allow developers to fully customise this or not? What's the best practice?

File: better-font-awesome.php [hunk 2]
```diff
@@ -513,18 +543,67 @@ public function text_callback( $args ) {
...
+		    <small>
+			    <a href="http://fontawesome.io/examples/">
+				    <?php echo esc_html_x( 'See all available options', 'For version 4.x +', 'better-font-awesome' ); ?> &raquo;
+			    </a>
+		    </small>
+		    <br/><br/>
+
+		    <?php
+		    $code_alternative_str = esc_html_x( 'or', 'Text between two variations of code markup examples', 'better-font-awesome' );
+		    //TODO: should I allow these example codes to be translated or not? Add a WP filter for them instead?
```
Comment (dashaluna) [inline better-font-awesome.php]: @tcrsavage also re: `coffee`, should I add a filter to allow developers to change what icon is displayed. Or be strict with it and just prescribe it?

File: better-font-awesome.php (hunk 1 above)
Comment (tcrsavage) [inline better-font-awesome.php]: If you can't picture reasonable use cases for it, I'd say no. I'm not a big fan of html filters. Would just ensure that the developer isn't blocked from extending the class and doing overrides that way for things like this.

File: better-font-awesome.php (hunk 2 above)
Comment (tcrsavage) [inline better-font-awesome.php]: I'd say no, don't bother

File: better-font-awesome.php [hunk 3]
```diff
@@ -513,18 +543,67 @@ public function text_callback( $args ) {
...
-                     <i class="icon-coffee fa fa-coffee icon-2x fa-2x"></i> <code>[icon name="coffee" class="icon-2x"]</code> or <code>&lt;i class="icon-coffee icon-2x"&gt;&lt;/i&gt;</code><br /><br />
-                     <i class="icon-coffee fa fa-coffee icon-2x fa-2x icon-rotate-90 fa-rotate-90"></i> <code>[icon name="coffee" class="icon-2x icon-rotate-90"]</code> or <code>&lt;i class="icon-coffee icon-2x icon-rotate-90"&gt;&lt;/i&gt;</code>',
-                'better-font-awesome' ) .
-                '</div>';
+
+	    ob_start();
+	    ?>
+
+	    <div class="bfa-usage-text">
+			<h3><?php esc_html_e( 'Usage', 'better-font-awesome' ); ?></h3>
```
Comment (tcrsavage) [inline better-font-awesome.php]: Indenting ;)

File: better-font-awesome.php [hunk 4]
```diff
@@ -513,18 +543,67 @@ public function text_callback( $args ) {
      * @return  string  Usage text.
      */
     public function get_usage_text() {
```
Comment (tcrsavage) [inline better-font-awesome.php:554]: So I mentioned templating. The idea there would be to do something like..

```
public function get_usage_text() { 
   ob_start();
   include __DIR__ . '/templates/usage-text.php'; 
   return ob_get_clean();
}
```

All your output would then be contained within usage-text.php

Makes readability a bit easier for other developers by keeping the more interesting handling separate from html output.

Comment (tcrsavage) [inline better-font-awesome.php:554]: On a 2nd note. get_usage_text() could be better named to something like `get_example_usage_html()`

File: better-font-awesome.php (hunk 1 above)
Comment (dashaluna) [inline better-font-awesome.php]: got you :+1:

File: better-font-awesome.php (hunk 4 above)
Comment (mattheu) [inline better-font-awesome.php:554]: or maybe even you could add something like this? Even cooler - you could pass args? 

``` php
function get_template( $template, $args ) {
    $path = apply_filters( 'bfa_get_template', __DIR__ . '/templates/usage-text.php', $args ); 
    if ( file_exists( $path ) ) {
        ob_start();
        include $path;
        return ob_get_clean();
    }
}
```

Updated to include filter as mentioned here: https://github.com/dashaluna/better-font-awesome/pull/2/files#r27222078

File: better-font-awesome.php (hunk 1 above)
Comment (mattheu) [inline better-font-awesome.php]: In fact - if you used a template function like listed above - you could make the template path filterable? That way devs could provide their own custom templates for things.

File: better-font-awesome.php [hunk 5]
```diff
@@ -453,26 +481,38 @@ public function version_callback( $versions ) {
...
             <p>
                 <?php 
-                printf( __( 'Version selection is currently unavailable. The attempt to reach the jsDelivr API server failed with the following error: %s', 'better-font-awesome' ), 
-                    '<code>' . $this->bfa_lib->get_error('api')->get_error_code() . ': ' . $this->bfa_lib->get_error('api')->get_error_message() . '</code>'
+                printf(
+	                esc_html__(
+		                'Version selection is currently unavailable. The attempt to reach the jsDelivr API server failed with the following error: %s',
+		                'better-font-awesome'
+	                ),
+	                '<code>' . $this->bfa_lib->get_error('api')->get_error_code() . ': ' . $this->bfa_lib->get_error('api')->get_error_message() . '</code>'
```
Comment (dashaluna) [inline better-font-awesome.php]: @tcrsavage do I need to `esc_html` this as in `'<code>' . esc_html( $this->bfa_lib->get_error('api')->get_error_code() ) .` ?

File: better-font-awesome.php (hunk 4 above)
Comment (dashaluna) [inline better-font-awesome.php:554]: @mattheu good point.. so making use of `$template`:

``` php
function get_template( $template, $args ) {
    $path = apply_filters( 'bfa_get_template', __DIR__ . '/templates/'. $template, $args ); 
    if ( file_exists( $path ) ) {
        ob_start();
        include $path;
        return ob_get_clean();
    }
}
```

Questions:
- I'm not sure what `$args` would do?
- Should I use `plugin_dir_path( __FILE__ )` instead of `__DIR__`?

File: better-font-awesome.php [hunk 6]
```diff
@@ -313,7 +336,7 @@ public function create_admin_page() {
     ?>
         <div class="wrap">
             <?php screen_icon(); ?>
```
Comment (pdewouters) [inline better-font-awesome.php:338]: this is deprecated

File: better-font-awesome.php (hunk 4 above)
Comment (mattheu) [inline better-font-awesome.php:554]: I guess we're going off on a bit of a tangent here...

When I've done something like this - I've added an args param that is an array of values that can be used in the template.

This is kinda similar to how you would use a templating like mustache. I think Mustache you pass JSON when rendering a template.

Probably over complicating things now!

Comment (mattheu) [inline better-font-awesome.php:554]: `plugin_dir_path( __FILE__ )` or `__DIR__`?

`__DIR__` is a shortcut for `dirname( __FILE__ )` but isn't PHP 5.2 compatable.

Internally `plugin_dir_path` just does `trailingslashit( dirname( __FILE__ ) );`

I'd probably use `plugin_dir_path` in a plugin - but often I do just use `__DIR__`.

File: better-font-awesome.php (hunk 6 above)
Comment (dashaluna) [inline better-font-awesome.php:338]: @pdewouters opened an issue :) thanks!

File: better-font-awesome.php (hunk 4 above)
Comment (tcrsavage) [inline better-font-awesome.php:554]: @matheu I advised to just use `__DIR__` in this case, the plugin was already using 5.3+ syntax. see `get_instance` method on this class.

File: better-font-awesome.php [hunk 7]
```diff
@@ -553,4 +583,28 @@ public function sanitize( $input ) {
...
+	/**
+	 * Outputs template's content
+	 *
+	 * @param string $template String name of a template file
+	 * @param array $args      Associative array that's accessible in the template via $args['key']
+	 *
+	 * @return string          Content of the template file if it exists
+	 *                         Empty string otherwise
+	 */
+	function get_template( $template, $args = array() ) {
```
Comment (tcrsavage) [inline better-font-awesome.php]: Always define access modifiers. This method can be `protected`.  `protected function get_template( $template, $args = array() )`

File: better-font-awesome.php [hunk 8]
```diff
@@ -335,25 +358,25 @@ public function create_admin_page() {
     function add_settings() {
```
Comment (tcrsavage) [inline better-font-awesome.php:358]: Access modifier

File: better-font-awesome.php [hunk 9]
```diff
@@ -392,18 +421,17 @@ function add_settings() {
      */
     function get_versions_list() {
```
Comment (tcrsavage) [inline better-font-awesome.php:416]: Access modifier

File: better-font-awesome.php [hunk 10]
```diff
@@ -382,30 +405,30 @@ function add_settings() {
...
-        }
+		} else {
+			$versions = array();
+		}
 
-        return $versions;
+		return $versions;
 
-    }
+	}
```
Comment (dashaluna) [inline better-font-awesome.php]: Grrrr, trying to revert the spacing alignments things - copied from raw original file, but that doesn't seem to work :( Any ideas or tips? @johnbillion

Comment (tcrsavage) [inline better-font-awesome.php]: @dashaluna you're on phpstorm right? Edit->Convert Indents

File: better-font-awesome.php (hunk 9 above)
Comment (dashaluna) [inline better-font-awesome.php:416]: @tcrsavage good stuff, will add it as an issue and will do a separate PR :)

File: better-font-awesome.php (hunk 5 above)
Comment (mattheu) [inline better-font-awesome.php]: Yes I think you should escape this.

File: better-font-awesome.php [hunk 11]
```diff
@@ -453,26 +476,38 @@ public function version_callback( $versions ) {
...
             <p>
                 <?php 
-                printf( __( 'Font Awesome will still render using version: %s', 'better-font-awesome' ),
-                    '<code>' . $this->bfa_lib->get_fallback_version() . '</code>'
+                printf(
+	                esc_html__(
+		                'Font Awesome will still render using version: %s',
+		                'better-font-awesome'
+	                ),
+	                '<code>' . $this->bfa_lib->get_fallback_version() . '</code>'
```
Comment (mattheu) [inline better-font-awesome.php]: And escape this.

File: better-font-awesome.php [hunk 12]
```diff
@@ -553,4 +582,28 @@ public function sanitize( $input ) {
 
     }
 
+	/**
```
Comment (tcrsavage) [inline better-font-awesome.php:585]: Indenting ;)
